src/log/log.c

```diff
@@ -462,6 +462,10 @@ __log_file_header(
 		WT_ERR(__log_acquire(session, logrec->len, &tmp));
 	}
 	WT_ERR(__log_fill(session, &myslot, 1, buf, NULL));
+	/*
+	 * Make sure the header gets to disk.
+	 */
+	WT_ERR(__wt_fsync(session, tmp.slot_fh));
```

Comment:
Well the comment is correct in that being called to write in the main log versus a pre-allocated log file it will be called with the lock.  Currently the issue you bring up was fixed in 662e26eeb31f76f2c4aeebf6.  We create the main log file into a temp name and only once it is created and the header written that we rename it into place.  So no, we should not have a window where there is a main log without a header.